gax-httpjson/src/test/java/com/google/api/gax/httpjson/testing/MockHttpService.java

```diff
@@ -0,0 +1,163 @@
+
+import com.google.api.client.http.LowLevelHttpRequest;
+import com.google.api.client.http.LowLevelHttpResponse;
+import com.google.api.client.testing.http.MockHttpTransport;
+import com.google.api.client.testing.http.MockLowLevelHttpRequest;
+import com.google.api.client.testing.http.MockLowLevelHttpResponse;
+import com.google.api.gax.httpjson.ApiMessage;
+import com.google.api.gax.httpjson.ApiMethodDescriptor;
+import com.google.api.pathtemplate.PathTemplate;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import java.io.StringWriter;
+import java.io.Writer;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Queue;
+
+/* Mocks an HTTPTransport. Expected responses and exceptions can be added to a queue
+ * from which this mock HttpTransport polls when it relays a response. */
+public final class MockHttpService extends MockHttpTransport {
+
+  private final List<String> requestPaths = new LinkedList<>();
+  private final Queue<HttpResponseFactory> responseHandlers = new LinkedList<>();
+  private List<ApiMethodDescriptor<? extends ApiMessage, ? extends ApiMessage>> serializers;
+  private String endpoint;
+
+  /* Create a MockHttpService.
+   *
+   * @param serializers - the list of method descriptors for the methods that the mocked server supports.
+   * @param fixedEndpoint - the fixed portion of the endpoint URL that prefixes the methods' path template substring. */
+  public MockHttpService(
+      List<ApiMethodDescriptor<? extends ApiMessage, ? extends ApiMessage>> serializers,
+      String fixedEndpoint) {
+    this.serializers = ImmutableList.copyOf(serializers);
+    this.endpoint = fixedEndpoint;
+  }
+
+  @Override
+  public LowLevelHttpRequest buildRequest(String method, final String url) {
+    requestPaths.add(url);
+    return new MockLowLevelHttpRequest() {
+      @Override
+      public LowLevelHttpResponse execute() {
+        return getHttpResponse(url);
+      }
+    };
+  }
+
+  /* Add an ApiMessage to the response queue. */
+  public void addResponse(final ApiMessage response) {
+    responseHandlers.add(
+        new HttpResponseFactory() {
+          @Override
+          public MockLowLevelHttpResponse getHttpResponse(String targetUrl) {
+            Writer writer = new StringWriter();
+            MockLowLevelHttpResponse httpResponse = new MockLowLevelHttpResponse();
+            Preconditions.checkArgument(
+                serializers != null, "MockHttpService has null serializers.");
+
+            // relativePath will be repeatedly truncated until it contains only
+            // the path template substring of the endpoint URL.
+            String relativePath = targetUrl.replaceFirst(endpoint, "");
+            int queryParamIndex = relativePath.indexOf("?");
+            queryParamIndex = queryParamIndex < 0 ? relativePath.length() : queryParamIndex;
+            relativePath = relativePath.substring(0, queryParamIndex);
```

Comment:
Maybe pull out the logic to get the relative path to shorten this method a little

Reply:
Done.